### PR TITLE
MINOR: A few cleanups of usage of Either in TC

### PR DESCRIPTION
**Pull request by hachikuji**

### CODE REVIEW

File: core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala [hunk 1]
```diff
@@ -213,36 +212,31 @@ class TransactionStateManager(brokerId: Int,
                                              createdTxnMetadata: Option[TransactionMetadata]): Either[Errors, Option[CoordinatorEpochAndTxnMetadata]] = {
     inReadLock(stateLock) {
       val partitionId = partitionFor(transactionalId)
-
       if (loadingPartitions.exists(_.txnPartitionId == partitionId))
-        return Left(Errors.COORDINATOR_LOAD_IN_PROGRESS)
-
-      if (leavingPartitions.exists(_.txnPartitionId == partitionId))
-        Right(Errors.NOT_COORDINATOR)
```
Comment (hachikuji) [inline core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala]: Note this was a bug spotted by @apurvam

**Comment by guozhangwang** [comment]
@hachikuji are the Jenkins failures relevant?

**Comment by hachikuji** [comment]
@guozhangwang Yes they are. I pushed a fix.

File: core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala [hunk 2]
```diff
@@ -199,8 +198,10 @@ class TransactionStateManager(brokerId: Int,
 
   def getTransactionState(transactionalId: String) = getAndMaybeAddTransactionState(transactionalId, None)
 
-  def putTransactionStateIfNotExists(transactionalId: String, txnMetadata: TransactionMetadata) =
```
Comment (ijuma) [inline core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala]: Can you please specify the result type of the public `getTransactionState` method as well?

File: core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala [hunk 3]
```diff
@@ -213,36 +214,30 @@ class TransactionStateManager(brokerId: Int,
...
+                    CoordinatorEpochAndTxnMetadata(cacheEntry.coordinatorEpoch, txnMetadata)
                   }
-              }
-
-            case currentTxnMetadata =>
-              Right(Some(CoordinatorEpochAndTxnMetadata(cacheEntry.coordinatorEpoch, currentTxnMetadata)))
-          }
+                }
+              case currentTxnMetadata =>
+                Some(CoordinatorEpochAndTxnMetadata(cacheEntry.coordinatorEpoch, currentTxnMetadata))
```
Comment (ijuma) [inline core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala]: Would it work to move the creation of this instance higher up and to just do the txn metadata within these two `case` statements?

**Review by ijuma — APPROVED** [review]
Thanks for the PR, this is a good improvement, it makes the code more idiomatic and avoids some bad patterns like matches that are not exhaustive. LGTM. Just one question and comment. No re-review needed.

File: core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala (hunk 3 above)
Comment (hachikuji) [reply core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala]: Probably not what you had in mind, but I pushed a simplification.

**Comment by hachikuji** [comment]
By the way, I'm going to hold off merging this until #3287 is in.

File: core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala [hunk 4]
```diff
@@ -221,19 +222,16 @@ class TransactionStateManager(brokerId: Int,
...
-              case currentTxnMetadata =>
-                Some(CoordinatorEpochAndTxnMetadata(cacheEntry.coordinatorEpoch, currentTxnMetadata))
-            })
+            val txnMetadata = Option(cacheEntry.metadataPerTransactionalId.get(transactionalId)).orElse {
+              createdTxnMetadata.map { txnMetadata =>
+                val currentTxnMetadata = cacheEntry.metadataPerTransactionalId.putIfNotExists(transactionalId, txnMetadata)
+                if (currentTxnMetadata != null)
+                  currentTxnMetadata
+                else
+                  txnMetadata
```
Comment (ijuma) [inline core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala]: You can simplify this one in a similar way as well:

```scala
Option(cacheEntry.metadataPerTransactionalId.putIfNotExists(transactionalId, txnMetadata)).getOrElse {
  txnMetadata
}
```

File: core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala (hunk 3 above)
Comment (ijuma) [reply core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala]: Your simplification looks good. It's roughly what I had in mind although I hadn't worked out all the details. I suggested one additional simplification.

File: core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala [hunk 5]
```diff
@@ -197,10 +196,13 @@ class TransactionStateManager(brokerId: Int,
...
 
-  def getTransactionState(transactionalId: String) = getAndMaybeAddTransactionState(transactionalId, None)
+  def getTransactionState(transactionalId: String): Either[Errors, Option[CoordinatorEpochAndTxnMetadata]] =
+    getAndMaybeAddTransactionState(transactionalId, None)
 
-  def putTransactionStateIfNotExists(transactionalId: String, txnMetadata: TransactionMetadata) =
+  def putTransactionStateIfNotExists(transactionalId: String,
+                                     txnMetadata: TransactionMetadata): Either[Errors, CoordinatorEpochAndTxnMetadata] =
     getAndMaybeAddTransactionState(transactionalId, Some(txnMetadata))
+      .right.map(_.getOrElse(throw new IllegalStateException("Expected non-empty transaction metadata returned")))
```
Comment (guozhangwang) [inline core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala]: nit: `Expected non-empty transaction metadata returned` reads a bit unclear. Were you trying to say `Unexpected empty transaction metadata returned while putting $txnMetadata`?

Comment (hachikuji) [reply core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala]: Ack

**Comment by guozhangwang** [comment]
One minor comment, otherwise LGTM!

**Comment by hachikuji** [comment]
Rebased against trunk. I will merge after the builds finish.

**Pull request closed without merging**